migrations/src/main/scala/za/co/absa/enceladus/migrations/framework/migration/CollectionMigration.scala

```diff
@@ -0,0 +1,293 @@
+
+  /** Returns a list of collections to be removed during the migration */
+  def getCollectionsToRemove: List[String] = collectionsToDrop.toList
+
+  /** Returns a list of collections to be removed during the migration */
+  def getCollectionsToRename: List[(String, String)] = collectionsToRename.toList
+
+  /** Returns a list of indexes to be added during the migration */
+  def getIndexesToAdd: List[Index] = indexesToCreate.toList
+
+  /** Returns a list of indexes to be removed during the migration */
+  def getIndexesToRemove: List[Index] = indexesToDrop.toList
+
+  /**
+    * If a migration adds or removes collections it should provide a new list of collections based
+    * on the list of collections available for he previous version of a database.
+    */
+  override def applyCollectionChanges(collections: List[String]): List[String] = {
+    var newCollections = collections
+    for (c <- getCollectionsToAdd) {
+      if (!newCollections.contains(c)) {
+        newCollections = newCollections :+ c
+      }
+    }
+    val collectionsToRemove = getCollectionsToRemove
+    newCollections = newCollections.filterNot(c => collectionsToRemove.contains(c))
+
+    val renameMap = collectionsToRename.toMap
+    newCollections.map(collectionName => {
+      renameMap.getOrElse(collectionName, collectionName)
+    })
+  }
+
+  /**
+    * Executes a migration on a given database and a list of collection names.
+    */
+  abstract override def execute(db: DocumentDb, collectionNames: Seq[String]): Unit = {
+    super.execute(db, collectionNames)
+    applyCreateCollections(db)
+    applyDropCollections(db)
+    applyRenameCollection(db)
+    applyIndexDrop(db)
+    applyIndexCreate(db)
+  }
+
+  /**
+    * Applies additions of new collections to the provided database.
+    */
+  private def applyCreateCollections(db: DocumentDb): Unit = {
+    collectionsToCreate.foreach(c => {
+      val newCollection = MigrationUtils.getVersionedCollectionName(c, targetVersion)
+      if (db.isCollectionExists(newCollection)) {
+        log.info(s"Dropping existing collection $newCollection")
+        db.dropCollection(newCollection)
+      }
+      log.info(s"Adding new collection $newCollection")
+      db.createCollection(newCollection)
+    })
+  }
+
+  /**
+    * Applies collection removals to the provided database.
+    */
+  private def applyDropCollections(db: DocumentDb): Unit = {
+    collectionsToDrop.foreach(c => {
+      val collection = MigrationUtils.getVersionedCollectionName(c, targetVersion)
+      log.info(s"Removing collection $collection")
+      db.dropCollection(collection)
+    })
+  }
+
+  /**
+    * Applies collection renames to the provided database.
+    */
+  private def applyRenameCollection(db: DocumentDb): Unit = {
+    collectionsToRename.foreach {
+      case (oldName, newName) =>
+        val oldVersionedCollection = MigrationUtils.getVersionedCollectionName(oldName, targetVersion)
```

Comment:
I don't very much get the use of this method. If I am doing migration and I want to rename something, is the target version going to be the same or -1?

Reply:
Migrator always clones something first, and after that applies all renames, drops, etc.